components/dashboard/src/settings/Integrations.tsx

```diff
@@ -199,106 +181,441 @@ function GitProviders() {
+        if (props.mode === "edit") {
+            setType(props.provider.type);
+            setHost(props.provider.host);
+            setClientId(props.provider.oauth.clientId);
+            setClientSecret(props.provider.oauth.clientSecret);
+            setRedirectURL(props.provider.oauth.callBackUrl);
+        }
+    }, []);
+
+    useEffect(() => {
+        validate();
+    }, [clientId, clientSecret])
+
+    const close = () => props.onClose && props.onClose();
+    const updateList = () => props.update && props.update();
+
+    const activate = async () => {
+        let entry = (props.mode === "new") ? {
+            host,
+            type,
+            clientId,
+            clientSecret,
+            ownerId: props.userId
+        } as AuthProviderEntry.NewEntry : {
+            id: props.provider.id,
+            ownerId: props.userId,
+            clientId,
+            clientSecret: clientSecret === "redacted" ? undefined : clientSecret
+        } as AuthProviderEntry.UpdateEntry;
+
+        setBusy(true);
+        setErrorMessage(undefined);
+        try {
+            const newProvider = await getGitpodService().server.updateOwnAuthProvider({ entry });
+
+            // the server is checking periodically for updates of dynamic providers, thus we need to
+            // wait at least 2 seconds for the changes to be propagated before we try to use this provider.
+            await new Promise(resolve => setTimeout(resolve, 2000));
+
+            updateList();
 
-    return <div className="flex mt-4">
-        <input className={"focus:ring-0 mt-1 rounded-sm cursor-pointer " + (props.checked ? 'bg-gray-800' : '')} type="checkbox"
-            id={checkboxId}
-            {...inputProps}
-        />
-        <div className="flex flex-col ml-2">
-            <label htmlFor={checkboxId} className="text-gray-700 text-md font-semibold">{props.title}</label>
-            <div className="text-gray-400 text-md">{props.desc}</div>
+            // just open the authorization window and do *not* await
+            openAuthorizeWindow({ host: newProvider.host, onSuccess: updateList });
+
+            // close the modal, as the creation phase is done anyways.
+            close();
+        } catch (error) {
+            console.log(error);
+            setErrorMessage("message" in error ? error.message : "Failed to update Git provider");
+        }
+        setBusy(false);
+    }
+
+    const updateHostValue = (host: string) => {
+        if (props.mode === "new") {
+            setHost(host);
+            setRedirectURL(callbackUrl(host));
+            setErrorMessage(undefined);
+        }
+    }
+
+    const updateClientId = (value: string) => {
+        setClientId(value);
+    }
+    const updateClientSecret = (value: string) => {
+        setClientSecret(value);
+    }
+
+    const validate = () => {
+        const errors: string[] = [];
+        if (clientId.trim().length === 0) {
+            errors.push("Client ID is missing.");
```

Comment:
**suggestion**: What do you think of hiding the error message when you are trying to add a new integration and showing this only after clicking _Activate Integration_?

Reply:
But you can no longer click that button, as both values are missing.

Reply:
Yes, but there's no need to also show the error dialog, right?